test/e2e/storage/pv_protection.go

```diff
@@ -129,3 +129,16 @@ var _ = utils.SIGDescribe("PV Protection", func() {
 		framework.WaitForPersistentVolumeDeleted(client, pv.Name, framework.Poll, framework.PVDeletingTimeout)
 	})
 })
+
+func checkForPVProtectionFinalizer(client clientset.Interface, pv *v1.PersistentVolume) error {
+	return wait.PollImmediate(time.Second, time.Minute, func() (bool, error) {
```

Comment:
For new PVs and PVCs, the finalizer is added by the [admission controller](https://sourcegraph.com/github.com/kubernetes/kubernetes/-/blob/plugin/pkg/admission/storage/storageobjectinuseprotection/admission.go#L84).  So I think it should automatically get added on creation and we shouldn't need to retry.

Reply:
@msau42 Thanks for your comment. You are right, the finalizer flag should be added automatically.
But the flag could be added a little lately after the volume creation, and if the e2e test runner is so faster than the test target k8s cluster (eg. with the separation between test runner and test target like my environment), the test runner reads the flag condition just before setting it automatically. Then the test can be failed.
Actually by applying this PR on my environment, the test result becomes stable and the retry happens once if this issue happens.

Reply:
The admission controller should add the finalizer before the API object gets committed to etcd.  Can you check that the admission controller is enabled in your environment?

Reply:
@msau42 Thanks for your comment. That is a question related to this. There is the similar code for adding the finalizder flag on https://github.com/kubernetes/kubernetes/blob/7b203c6809a5af940dbd8962b2dc65f93b33b401/pkg/controller/volume/pvprotection/pv_protection_controller.go#L157
Can I know what is difference between them?
Anyways I will check my environment soon.

Reply:
The logic in the protection controller itself is to handle old PVs that have been upgraded.  All new PVs should be handled by the admission controller.

Reply:
As you guessed, my environment doesn't seem to have the admission controller like
```
$ kubectl get pods --all-namespaces

NAMESPACE         NAME                                             READY   STATUS    RESTARTS   AGE
default           standalone-cinder-provisioner-75548544b4-89v75   1/1     Running   1          52d
kube-system       coredns-86c58d9df4-rv4x4                         1/1     Running   0          41d
kube-system       coredns-86c58d9df4-zqlmv                         1/1     Running   0          41d
kube-system       etcd-k8s-master                                  1/1     Running   0          63d
kube-system       kube-apiserver-k8s-master                        1/1     Running   0          63d
kube-system       kube-controller-manager-k8s-master               1/1     Running   0          42d
kube-system       kube-flannel-ds-amd64-5wqtz                      1/1     Running   0          63d
kube-system       kube-flannel-ds-amd64-d4n8n                      1/1     Running   0          41d
kube-system       kube-flannel-ds-amd64-l64rw                      1/1     Running   1          52d
kube-system       kube-proxy-fgmlq                                 1/1     Running   1          63d
kube-system       kube-proxy-gshbj                                 1/1     Running   0          63d
kube-system       kube-proxy-qxtrj                                 1/1     Running   1          63d
kube-system       kube-scheduler-k8s-master                        1/1     Running   0          53d
kube-system       metrics-server-68d85f76bb-59nqw                  1/1     Running   1          52d
```
But the finalizer flag is added after the creation as I said.
So I'd like to understand the above difference.

Reply:
The admission controller is part of the API server process and has to be enabled via flags to the apiserver.  Here is mine:
```
/usr/local/bin/kube-apiserver <snip> --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,PodPreset,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,Priority,StorageObjectInUseProtection,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota --min-request-timeout=300 --runtime-config=batch/v2alpha1=true --feature-gates=ExperimentalCriticalPodAnnotation=true --advertise-address=130.211.208.186 --authorization-mode=Node,RBAC --allow-privileged=true
```

Reply:
Thanks again.
I checked the log of apiserver and the loaded admission controllers seem like
```
plugins.go:161] Loaded 6 validating admission controller(s) successfully in the following order: LimitRanger,ServiceAccount,Pri
ority,PersistentVolumeClaimResize,ValidatingAdmissionWebhook,ResourceQuota.
```

Reply:
You will need to add: StorageObjectInUseProtection

Reply:
`NamespaceLifecycle, PersistentVolumeLabel, PodPreset,DefaultStorageClass, DefaultTolerationSeconds, NodeRestriction, StorageObjectInUseProtection, MutatingAdmissionWebhook` seem lack of my environment by comparing yours.

Reply:
@msau42 Thank you very much for your advice!

Reply:
Ref: https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#storage-object-in-use-protection
```
The StorageObjectInUseProtection plugin adds the kubernetes.io/pvc-protection or kubernetes.io/pv-protection finalizers to newly created Persistent Volume Claims (PVCs) or Persistent Volumes (PV).
```

Reply:
drat. good catch @msau42 . I totally forgot about the admission controller. :(